src/Mews.Fiscalization.Core/Model/Collections/Sequence.cs

```diff
@@ -0,0 +1,81 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using FuncSharp;
+
+namespace Mews.Fiscalization.Core.Model
+{
+    public interface ISequence<T> : IReadOnlyList<Indexed<T>>
+    {
+        INonEmptyEnumerable<Indexed<T>> Values { get; }
+
+        int StartIndex { get; }
+    }
+
+    public class Sequence<T> : ISequence<T>
+    {
+        private Sequence(INonEmptyEnumerable<Indexed<T>> values)
+        {
+            Check.Condition(values.IsSequential(v => v.Index, startIndex: values.First().Index), "Item indexes are not sequential.");
+            Values = values;
+        }
+
+        public INonEmptyEnumerable<Indexed<T>> Values { get; }
+
+        public int StartIndex => Values.First().Index;
```

Comment:
```suggestion
        public int StartIndex => Values.Head.Index;
```

Nit: Using the `Head` everywhere is nicer.